go-controller/pkg/cni/cniserver.go

```diff
@@ -155,6 +212,13 @@ func (s *Server) handleCNIRequest(w http.ResponseWriter, r *http.Request) {
 			klog.Warningf("Error writing %s HTTP response: %v", req.Command, err)
 		}
 	}
+
+	if req.Command == CNIAdd {
```

Comment:
should the CNIDel  clear the cache too?

Reply:
or cancel the pending Adds, if they exist?

Reply:
we cant get a CNIDel for a CNIAdd that is already happening, I believe those events are serial

Reply:
A CNI ADD and DEL for the *same sandbox* are always serialized. We do not care about DEL for two reasons: (1) Delete doesn't touch anything that other sandboxes would care about, and (2) DEL should clean up as much as it can. So we can (and should) just let delete run since it won't affect anything else, and we should let delete run because it still needs to clean up stuff that a completed ADD may have allocated. If we canceled the DEL, it might stop before cleaning everything up.

Reply:
Also, we only want to clean up this specific sandbox request, not one for another sandbox of the same namespace/name since htat one might be the active one.

Reply:
Would it make sense to call cancelPodRequest for the previous ADD when the cniAdd happens for the same `namespace/name` again? If so, can we then get rid of the timestamp logic altogether since there will only be one pending add at any point.

Reply:
It might entail caching the sandbox-id of the active request in a map[namespace\name]->sandbox_id

Reply:
can we add a comment explaining this too?
https://github.com/ovn-org/ovn-kubernetes/pull/1830/files#r522996143

Reply:
> Would it make sense to call cancelPodRequest for the previous ADD when the cniAdd happens for the same `namespace/name` again? If so, can we then get rid of the timestamp logic altogether since there will only be one pending add at any point.

@abhat yes, this makes sense, but we can't do this without keeping sandboxID as the map key I think. Otherwise removing from the map becomes more complicated if the request got canceled vs. finished normally.

Reply:
Well, I didn't mean to get rid of the logic when the informer cache del happens when DeleteFunc is called. I am saying we can simplify the deletion logic by not relying on timestamps and instead assume that there will be only one outstanding request at any point. If the second add happens, we can cancel the pending add. So in some ways, there are two possibilities for canceling the pod request - a) pod del update comes via the informer cache update b) a cniAdd happens for the same pod (`namespace/name`). 

When the request finishes normally, you have all the necessary data to remove from the map at hand right? When the request gets cancelled via the informer del, you still only need to know the `namespace\name` for the pod if that is the key into the map. Isn't that right?

Reply:
@abhat not quite. The problem I see is that the existing sandbox request would need to be cleared from two places: (1) at the end of handleCNIRequest() when the request has finished normally, and (2) a sub-function of handleCNIRequest() when a newer request comes in. These two things are in two different goroutines though.

One way I can think of is to check if the map value is the same request object as is being removed:

```
func makeRequestKey(namespace, name string) string {
	return req.PodNamespace + "/" + req.PodName)
}

func (s *Server) startSandboxRequest(req *PodRequest) error {
	// Only sandbox add requests are tracked because only adds need
	// to be canceled when a newer request comes in.
	if req.Command == CNIAdd {
		s.runningSandboxAddsLock.Lock()
		defer s.runningSandboxAddsLock.Unlock()
		key := makeRequestKey(req.PodNamespace, req.PodName)
		if _, ok := s.runningSandboxAdds[key]; ok {
			req.cancel()
		}
		s.runningSandboxAdds[key] = req
	}
	return nil
}

func (s *Server) finishSandboxRequest(req *PodRequest) {
	if req.Command == CNIAdd {
		s.runningSandboxAddsLock.Lock()
		defer s.runningSandboxAddsLock.Unlock()
		key := makeRequestKey(req.PodNamespace, req.PodName)
		if existing, ok := s.runningSandboxAdds[key]; ok && existing == req {
			delete(s.runningSandboxAdds, key)
		}
	}
}
```

but that seems pretty ugly and is not a pattern I've commonly seen.

Reply:
You can simplify it further - pass the `cancel` vs `delete` request semantics as an argument in the two different scenarios.

Reply:
And for the record, this doesn't work if we get a ADD1, ADD2, Pod1 Delete as the ADD2 would have canceled the ADD1, and then the Pod1 Delete would cancel the ADD2 which is still valid.

Reply:
if we can get the sandbox id when Pod1 Delete happens and if it doesn't match the ADD2's sandbox-id, we can skip?

Reply:
@abhat Pod1 Delete would be an api event, not a CNI event, so we don't get the sandbox ID with it.